### PR TITLE
Implement BananaButtonStateProvider

### PR DESCRIPTION
**Pull request by jdragonbae**
This patch implements BananaButtonStateProvider to get state of buttons
such as download, find in page, add to home and share.

Refs: #1029

### CODE REVIEW

File: src/org/triple/banana/quick_menu/ViewControllerImpl.java [hunk 1]
```diff
@@ -46,16 +46,39 @@ public boolean onClickQuickMenuButton(@IdRes int buttonId) {
...
+            label = R.string.download;
+        } else if (id == R.id.find_in_page) {
+            image = R.drawable.ic_find_in_page_black_24dp;
+            label = R.string.find_in_page;
+        } else if (id == R.id.share) {
+            image = R.drawable.ic_share_black_24dp;
+            label = R.string.share;
+        } else if (id == R.id.add_to_home) {
+            image = R.drawable.ic_add_to_home_black_24dp;
+            label = R.string.add_to_home;
```
Comment (romandev) [inline src/org/triple/banana/quick_menu/ViewControllerImpl.java:72]:
```
boolean enabled = true;
switch (state) {
    case ButtonState.MOBILE_PAGE:
        image = ...;
        break;
    case ButtonState.DESKTOP_PAGE:
        image = ...;
        break;
    case ButtonState.DISABLE:
        enabled = false;
        break;
    case ButtonState.ENABLE:
        // The enabled is already `true`. So, do nothing here.
        break;
}
```

Also, you should update `updateButton` like this:
```
if (newInfo.image != 0) {
    exitingInfo.image = newInfo.image;
}
if (newInfo.label != 0) {
    existingInfo.label = newInfo.label;
}
```

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaButtonStateProvider.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+package org.banana.cake.interfaces;
+
+import androidx.annotation.NonNull;
+
+public interface BananaButtonStateProvider {
+    static @NonNull BananaButtonStateProvider get() {
+        return BananaInterfaceProvider.get(BananaButtonStateProvider.class);
+    }
+
+    boolean isEnableDownloadPage();
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply banana_cake/interfaces/java/org/banana/cake/interfaces/BananaButtonStateProvider.java:15]: Done

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaButtonStateProvider.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    boolean isEnableShare();
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply banana_cake/interfaces/java/org/banana/cake/interfaces/BananaButtonStateProvider.java:16]: Done

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaButtonStateProvider.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    boolean isEnableFindInPage();
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply banana_cake/interfaces/java/org/banana/cake/interfaces/BananaButtonStateProvider.java:17]: Done

File: banana_cake/interfaces/java/org/banana/cake/interfaces/BananaButtonStateProvider.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    boolean isEnableAddToHome();
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply banana_cake/interfaces/java/org/banana/cake/interfaces/BananaButtonStateProvider.java:18]: Done

File: banana_cake/java/org/banana/cake/CakeButtonStateProvider.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    public boolean isEnableShare() {
+        Tab currentTab = getActivityTab();
+        if (currentTab == null) {
+            return false;
+        }
+        String url = currentTab.getUrlString();
+        boolean isChromeScheme = url.startsWith(UrlConstants.CHROME_URL_PREFIX)
+                || url.startsWith(UrlConstants.CHROME_NATIVE_URL_PREFIX);
+        boolean isDataScheme = url.startsWith(UrlConstants.DATA_SCHEME);
+        return !isChromeScheme && !isDataScheme;
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply banana_cake/java/org/banana/cake/CakeButtonStateProvider.java:42]: Thanks for the your guidance. I'll try to use `shouldEnableShare`.

File: src/org/triple/banana/quick_menu/ButtonInfo.java [hunk 7]
```diff
@@ -16,10 +16,12 @@
     int image;
     @StringRes
     int label;
+    boolean enabled;
 
-    ButtonInfo(@IdRes int id, @IdRes int image, @StringRes int label) {
+    ButtonInfo(@IdRes int id, @IdRes int image, @StringRes int label, boolean enabled) {
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/quick_menu/ButtonInfo.java:21]: Done

File: src/org/triple/banana/quick_menu/ViewControllerImpl.java (hunk 1 above)
Comment (jdragonbae) [reply src/org/triple/banana/quick_menu/ViewControllerImpl.java:72]: SGTM. It seems to be more simpler. Thanks for the review.

File: src/org/triple/banana/quick_menu/QuickMenuMiddleButton.java [hunk 8]
```diff
@@ -55,7 +55,9 @@ public void setCompoundDrawablesWithIntrinsicBounds(@Nullable Drawable left,
         final int paddingSize =
                 getResources().getDimensionPixelSize(R.dimen.quick_menu_button_background_padding);
         iconDrawable.setLayerInset(1, paddingSize, paddingSize, paddingSize, paddingSize);
-
+        if (!isEnabled()) {
+            setAlpha(0.25f);
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/quick_menu/QuickMenuMiddleButton.java:67]: As we discussed via slack, move this code to the `setEnabled` overridden.

[the remark this answers is not in the document]
Comment (romandev) [reply src/org/triple/banana/quick_menu/QuickMenuMiddleButton.java:67]: BTW, if the button is enabled again, then shouldn't we update the alpha value?

[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/quick_menu/QuickMenuMiddleButton.java:69]: Thanks for the review. You're right. I'll update it.

**Review by romandev — APPROVED** [review]
lgtm

**Pull request merged**